spec.emu

```diff
@@ -40,412 +40,48 @@ contributors: Surma, Dan Ehrenberg
-  <p>The <dfn>ModuleBlock prototype object</dfn>:</p>
-  <ul>
-    <li>is <dfn>%ModuleBlock.prototype%</dfn>.</li>
-    <li>has a [[Prototype]] internal slot whose value is %Object.prototype%.</li>
-    <li>is an ordinary object.</li>
-  </ul>
-
-  <emu-clause id="sec-moduleblock.prototype.tostring">
-    <h1>ModuleBlock.prototype.toString ( )</h1>
-    <p>The following steps are taken:</p>
-    <emu-alg>
-      1. Let _module_ be the *this* value.
-      1. If Type(_module_) is Object and has the internal slot [[ModuleBlockBody]]
-        1. Return _module_.[[SourceText]].
-      1. Throw a *TypeError* exception.
-    </emu-alg>
-  </emu-clause>
-</emu-clause>
-
-<emu-clause id="sec-import-calls">
-  <h1>Import Calls</h1>
-
-  <emu-clause id="sec-import-call-runtime-semantics-evaluation">
-    <h1>Runtime Semantics: Evaluation</h1>
-
-    <emu-grammar>ImportCall : `import` `(` AssignmentExpression `)`</emu-grammar>
     <emu-alg>
       1. Let _referencingScriptOrModule_ be GetActiveScriptOrModule().
-      1. Let _argRef_ be the result of evaluating |AssignmentExpression|.
-      1. Let _specifier_ be ? GetValue(_argRef_).
-      1. Let _promiseCapability_ be ! NewPromiseCapability(%Promise%).
-      1. <del>Let _specifierString_ be ToString(_specifier_).</del>
-      1. <ins>If Type(_specifier_) is not Object or _specifier_ does not have the internal slot [[ModuleBlockBody]]</ins>
-        1. <ins>Set _specifier_ to ToString(_specifier_).</ins>
-        1. IfAbruptRejectPromise(<del>_specifierString_</del><ins>_specifier_</ins>, _promiseCapability_).
-      1. Perform ! <del>HostImportModuleDynamically</del><ins>ImportModuleDynamically</ins>(_referencingScriptOrModule_, <del>_specifierString_</del><ins>_specifier_</ins>, _promiseCapability_).
-      1. Return _promiseCapability_.[[Promise]].
+      1. Let _moduleSource_ be ? OrdinaryObjectCreate(*"%ModuleSource.prototype%"*, « [[ModuleSource]] »).
+      1. Set _moduleSource_.[[ModuleSource]] to ? CreateModuleSourceRecord(_body_).
+      1. Let _O_ be ? OrdinaryObjectCreate(*"%Module.prototype%"*, « [[Module]], [[ModuleSourceInstance]], [[ImportHook]] »).
+      1. Let _moduleRecord_ to ! CreateModuleRecord(_moduleSource_.[[ModuleSource]]).
+      1. Set _moduleRecord_.[[ModuleInstance]] to _O_.
+      1. Set _O_.[[Module]] to _moduleRecord_.
+      1. Set _O_.[[ModuleSourceInstance]] to _moduleSource_.
+      1. If _referencingScriptOrModule_ is not a Module Record or its [[ModuleInstance]] internal slot value is *undefined*, then
+        1. Let _importHookClosure_ be a new Abstract Closure with parameters (_specifier_) that captures _referencingScriptOrModule_ and performs the following steps when called
+          1. Assert: Type(_specifier_) is String.
+          1. Let _promiseCapability_ be ! NewPromiseCapability(%Promise%).
+          1. Perform HostImportModuleDynamically(_referencingScriptOrModule_, _specifier_, _promiseCapability_).
```

Comment:
Instead of passing `_referencingScriptOrModule_`, we might consider passing `_moduleRecord_`. This gives less strong guarantees (hosts could make the two imports in `import "a"; module { import "a"  }` resolve to different modules), but it makes it easier to serialize module blocks that have the default import hook (since they don't capture the parent module).